### PR TITLE
internal/ethapi: implement eth_getBlockReceipts

### PR DESCRIPTION
**Pull request by jsvisa**
implment https://github.com/ethereum/execution-apis/pull/438

### CODE REVIEW

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -897,6 +897,43 @@ func (s *BlockChainAPI) GetStorageAt(ctx context.Context, address common.Address
...
+		hash = h
+	} else {
+		block, err := s.b.BlockByNumberOrHash(ctx, blockNrOrHash)
+		if err != nil {
+			return nil, err
+		}
+		hash = block.Hash()
+	}
+
+	block, err := s.b.BlockByHash(ctx, hash)
```
Comment (s1na) [inline internal/ethapi/api.go:913]: Why not directly use `s.b.BlockByNumberOrHash`? IMO not worth the extra 10 lines to save on hashing a block.

**Comment by s1na** [comment]
Also please add to `internal/web3ext/web3ext.go`.

File: internal/ethapi/api.go (hunk 1 above)
Comment (jsvisa) [reply internal/ethapi/api.go:913]: thanks for this 👍

**Comment by jsvisa** [comment]
We need to add test cases, this depends on https://github.com/ethereum/go-ethereum/pull/27527, so I'll mark this PR ready after #27527 is merged.

**Comment by jsvisa** [comment]
@s1na @lightclient PTAL

File: internal/ethapi/api.go [hunk 2]
```diff
@@ -897,6 +897,34 @@ func (s *BlockChainAPI) GetStorageAt(ctx context.Context, address common.Address
...
+		// as per specification.
+		return nil, nil
+	}
+	receipts, err := s.b.GetReceipts(ctx, block.Hash())
+	if err != nil {
+		return nil, err
+	}
+	txs := block.Transactions()
+	if len(txs) != len(receipts) {
+		return nil, fmt.Errorf("receipts length mismatch: %d vs %d", len(txs), len(receipts))
```
Comment (s1na) [inline internal/ethapi/api.go:914]:
```suggestion
		return nil, fmt.Errorf("receipts length mismatch. want: %d, have: %d", len(txs), len(receipts))
```

File: internal/jsre/deps/web3.js [hunk 3]
```diff
@@ -3706,6 +3706,13 @@ var inputDefaultBlockNumberFormatter = function (blockNumber) {
     return inputBlockNumberFormatter(blockNumber);
 };
 
+var inputBlockNumberOrHashFormatter = function (numberOrHash) {
```
Comment (s1na) [inline internal/jsre/deps/web3.js:3709]: Yeah not sure about changing the dependency library directly.

File: internal/ethapi/api_test.go [hunk 4]
```diff
@@ -1140,3 +1137,102 @@ func TestRPCGetTransactionReceipt(t *testing.T) {
 		require.JSONEqf(t, want, have, "test %d: json not match, want: %s, have: %s", i, want, have)
 	}
 }
+
+func TestRPCGetBlockReceipts(t *testing.T) {
```
Comment (s1na) [inline internal/ethapi/api_test.go:1141]: Great thx for tests 👍

File: internal/ethapi/api.go [hunk 5]
```diff
@@ -1701,13 +1729,18 @@ func (s *TransactionAPI) GetTransactionReceipt(ctx context.Context, hash common.
 
 	// Derive the sender.
 	signer := types.MakeSigner(s.b.ChainConfig(), header.Number, header.Time)
+	return marshalReceipt(receipt, blockHash, blockNumber, signer, tx, int(index)), nil
+}
+
+// marshalReceipt marshals a transaction receipt into a JSON object.
+func marshalReceipt(receipt *types.Receipt, blockHash common.Hash, blockNumber uint64, signer types.Signer, tx *types.Transaction, txIndex int) map[string]interface{} {
```
Comment (s1na) [inline internal/ethapi/api.go:1736]: I think we're using interfaces as response objects too much in this package. You don't need to worry about it in this PR but generally pointing out that it'd be nice to use proper types more.

Comment (s1na) [reply internal/ethapi/api.go:1736]: Please ignore this as per https://github.com/ethereum/go-ethereum/pull/27305#issuecomment-1660277875

Comment (jsvisa) [reply internal/ethapi/api.go:1736]: sgtm

File: internal/jsre/deps/web3.js (hunk 3 above)
Comment (jsvisa) [reply internal/jsre/deps/web3.js:3709]: Which one will succeed, ethers.js, web3.js or another package?

Comment (s1na) [reply internal/jsre/deps/web3.js:3709]: We should just not directly change a 3rd party library in our code.

As for your question: we're now at the latest version of web3 that doesn't use Promises and is sync. We'd like to maintain the ease of working with sync methods and chaining them etc.
It'd be nice to upgrade to ethers.js at some point if we can make it seamless for users. Our JS vm (goja) has support for promises and async/await. However ideally we should avoid for users to do promise handling in console all the time.

**Review by s1na — APPROVED** [review]
LGTM thanks!

File: internal/ethapi/api.go (hunk 2 above)
Comment (holiman) [inline internal/ethapi/api.go:914]: Why is this check needed? I mean, if this were ever to be triggered, it basically means there is a very strange data corruption that has occurred, right? 

If so, I think it's overly defensive to have this check here, unless there's some compelling reason for it

**Review by holiman — COMMENTED** [review]
LGTM, minor nit/question only

File: internal/ethapi/api.go (hunk 2 above)
Comment (jsvisa) [reply internal/ethapi/api.go:914]: As a guarantee of data integrity, is that too conservative?

Comment (holiman) [reply internal/ethapi/api.go:914]: Yeah ok, that is an acceptable reason to me

**Review by holiman — APPROVED** [review]
LGTM

**Comment by holiman** [comment]
Should we wait for https://github.com/ethereum/execution-apis/pull/438 before merging? @lightclient what do  you say?

**Comment by hitdavid** [comment]
Will this PR be merged into the next release please?
Thanks

**Pull request merged**